lib/raii.md

```diff
@@ -0,0 +1,152 @@
+```
+If the right `__copy` method is not implemented but a user defined `__cast` metamethod exists that can cast one of the arguments to the correct type, then the cast is performed and then the relevant copy method is applied.
+
+### Object destruction
+`__dtor` can be used to free heap memory
+```
+    A.methods.__dtor(self : &A)
+```
+The implementation adds a deferred call to `__dtor ` near the end of a scope, right before a potential return statement, for all variables local to the current scope that are not returned. Hence, `__dtor` is tied to the lifetime of the object. For example, for a block of code the compiler would generate
+```
+do
+    var x : A, y : A
+    ...
+    ...
+    defer x:__dtor()    --generated by compiler
+    defer y:__dtor()    --generated by compiler
+end
+```
+or in case of a terra function
+```
+terra foo(x : A)
+    var y : A, z : A
+    ...
+    ...
+    defer z:__dtor()    --generated by compiler
+    return y
+end
+```
+`__dtor` is also called before any regular assignment (if a __copy method is not implemented)  to free 'old' resources. So
+```
+    a = b
+```
+is replaced by
+```
+    a:__dtor()  --generated by compiler
+    a = b
+```
+## Compositional API's
+If a struct has fields or subfields that are managed types, but do not implement `__init`, `__copy` or `__dtor`, then the compiler will generate default methods that inductively call existing `__init`, `__copy` or `__dtor` methods for its fields and subfields. This enables compositional API's like `vector(vector(int))` or  `vector(string)`. This is implemented as an extension to *terralib.lua* in *lib/terralibext.t*.
+
+## Examples
+The following files have been added to the terra testsuite:
+* *raii.t* tests whether `__dtor`, `__init`, and `__copy` are evaluated correctly for simple datatypes.
+* *raii-copyctr-cast.t* tests the combination of `metamethods.__cast` and `__copy`.
+* *raii-unique_ptr.t* tests some functionality of a unique pointer type.
+* *raii-shared_ptr.t* tests some functionality of a shared pointer type.
+* *raii-offset_ptr.t* tests some functionality of an offset pointer implementation, found e.g. in *boost.cpp*.
+* *raii-compose.t* tests the compositional aspect.
+
+You can have a look there for some common code patterns.
+
+## Current limitations
+* The implementation is not aware of when an actual heap allocation is made and therefore assumes that a managed variable always carries a heap resource. It is up to the programmer to properly initialize pointer variables to nil to avoid calling 'free' on uninitialized pointers.
+* Tuple (copy) assignment (regular or using `__copy`) are prohibited by the compiler in case of managed variables. This is done to prevent memory leaks or unwanted deletions in assignments such as
+```
+    a, b = b, a
+```
+* Currently, there is no way to prevent unwanted calls to `__dtor` in cases such as the following. Consider
+```
+terra foo()
+    var b : A
+    return bar(b)
```

Comment:
This appears to be pass-by-value so is `b` not copied on being passed to `bar`?

Reply:
Yes, but terra performs shallow copies. Indeed, the example would also hold if passed by reference.

Honestly, it's a very bad code pattern that should not compile. I've removed it from the documentation. I'll add a section later on "don'ts"